block-storage/src/main/scala/coop/rchain/blockstorage/dag/PersistentIndex.scala

```diff
@@ -0,0 +1,189 @@
+    } yield ()
+
+  def add(key: K, value: V): F[Unit] =
+    for {
+      toAppend <- keyValueCodec.encode((key, value)) match {
+                   case Successful(encoded) => encoded.toByteArray.pure[F]
+                   case Failure(e)          => Sync[F].raiseError[Array[Byte]](EncodeError(e))
+                 }
+      _ <- outputStream.write(toAppend)
+      _ <- outputStream.flush
+      _ <- crc.update(toAppend).flatMap(_ => updateCrcFile)
+      _ <- persistedDataState.modify(_.updated(key, value))
+    } yield ()
+
+  def addAll(newData: List[(K, V)]) =
+    newData.traverse_(add)
+
+  def data: F[Map[K, V]] =
+    persistedDataState.get
+
+  def close: F[Unit] =
+    outputStream.close
+}
+
+object PersistentIndex {
+  private final case class EncodeError(e: Err) extends Exception {
+    override def getMessage: String = s"Could not encode: $e"
+  }
+
+  private final case class DecodeError(e: Err) extends Exception {
+    override def getMessage: String = s"Could not decode: $e"
+  }
+
+  private def truncateDataLog[F[_]: Sync, K, V](
+      randomAccessIO: RandomAccessIO[F],
+      dataCollection: List[(K, V)],
+      keyValueCodec: Codec[(K, V)]
+  ): F[Unit] = {
+    val lastEntry = dataCollection.last
+    for {
+      lastEntrySize <- keyValueCodec.encode(lastEntry) match {
+                        case Successful(encoded) => ((encoded.size + 7) / 8).pure[F]
+                        case Failure(e) =>
+                          Sync[F].raiseError[Long](EncodeError(e))
+                      }
+      length <- randomAccessIO.length
+      _      <- randomAccessIO.setLength(length - lastEntrySize)
+    } yield ()
+  }
+
+  private def readCrc[F[_]: Sync: Log: RaiseIOError](crcPath: Path): F[Long] =
+    for {
+      _          <- createNewFile[F](crcPath)
+      bytes      <- readAllBytesFromFile[F](crcPath)
+      byteBuffer = ByteBuffer.wrap(bytes)
+      result <- Sync[F].delay { byteBuffer.getLong() }.handleErrorWith {
+                 case _: BufferUnderflowException =>
+                   for {
+                     _ <- Log[F].warn(s"CRC file $crcPath did not contain a valid CRC value")
+                   } yield 0
+                 case exception =>
+                   Sync[F].raiseError(exception)
+               }
+    } yield result
+
+  private def calculateCrc[F[_]: Sync, K, V](
+      dataCollection: List[(K, V)],
+      keyValueCodec: Codec[(K, V)]
+  ): F[Crc32[F]] = {
+    val crc = Crc32.empty[F]()
+    dataCollection.traverse_[F, Unit] {
+      case (key, value) =>
+        keyValueCodec.encode((key, value)) match {
+          case Successful(encoded) =>
+            crc.update(encoded.toByteArray)
+          case Failure(e) =>
+            Sync[F].raiseError(EncodeError(e))
+        }
+    } >> crc.pure[F]
+  }
+
+  def readDataList[F[_]: Sync, K, V](
+      dataPath: Path,
+      keyValueCodec: Codec[(K, V)],
+      corruptedError: => StorageError
+  ): F[List[(K, V)]] = {
```

Comment:
this will run the 
```
val inputStream = Files.newInputStream(dataPath, StandardOpenOption.READ)
    val bitVector   = BitVector.fromInputStream(inputStream)
```
on invocation.
I don't think it's the intended behaviour?

Reply:
Yep, totally forgot to rewrite this part after prototyping